File: Examples/Conversations/Obj-C/ConversationsExample/QuestionsViewController.m
//
//  QuestionsViewController.m
//  ConversationsExample
//
//  Copyright © 2016 Bazaarvoice. All rights reserved.
//

#import "QuestionsViewController.h"
#import "AnswersViewController.h"
#import "MyQuestionTableViewCell.h"

@import BVSDK;

@interface QuestionsViewController () <UITableViewDataSource,
                                       UITableViewDelegate>

@property(weak, nonatomic) IBOutlet BVQuestionsTableView *questionsTableView;

@property NSArray<BVQuestion *> *questions;

@end

@implementation QuestionsViewController

- (void)viewDidLoad {
  [super viewDidLoad];

  self.questions = [NSArray array];

  self.questionsTableView.dataSource = self;
  self.questionsTableView.estimatedRowHeight = 44;
  self.questionsTableView.rowHeight = UITableViewAutomaticDimension;
  [self.questionsTableView
                 registerNib:[UINib nibWithNibName:@"MyQuestionTableViewCell"
                                            bundle:nil]
      forCellReuseIdentifier:@"MyQuestionTableViewCell"];

  BVQuestionsAndAnswersRequest *request =
      [[BVQuestionsAndAnswersRequest alloc] initWithProductId:@"test1"
                                                        limit:20
                                                       offset:0];
  // optionally add in a sort option
  [request addQuestionSort:BVSortOptionQuestionsLastModeratedTime
                     order:BVSortOrderDescending];
  // optionally add in a filter
  [request addFilter:BVQuestionFilterTypeHasAnswers
      filterOperator:BVFilterOperatorEqualTo
               value:@"true"];

  [self.questionsTableView load:request
      success:^(BVQuestionsAndAnswersResponse *_Nonnull response) {
        self.questions = response.results;
        [self.questionsTableView reloadData];
      }
      failure:^(NSArray<NSError *> *_Nonnull errors) {
        NSLog(@"Error loading quesitons");
      }];
}

#pragma mark UITableViewDatasource

- (NSString *)tableView:(UITableView *)tableView
    titleForHeaderInSection:(NSInteger)section {
  return @"Questions Responses";
}

- (NSInteger)tableView:(UITableView *)tableView
    numberOfRowsInSection:(NSInteger)section {
  return [self.questions count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath {

  MyQuestionTableViewCell *cell =
      [tableView dequeueReusableCellWithIdentifier:@"MyQuestionTableViewCell"];

  cell.question = [self.questions objectAtIndex:indexPath.row];

  return cell;
}

#pragma mark UITableViewDelegate

- (void)tableView:(UITableView *)tableView
    didSelectRowAtIndexPath:(NSIndexPath *)indexPath {

  BVQuestion *question = [self.questions objectAtIndex:indexPath.row];

  UIStoryboard *sb = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
  AnswersViewController *answersVC =
      [sb instantiateViewControllerWithIdentifier:@"AnswersViewController"];
  answersVC.question = question;

  [self.navigationController pushViewController:answersVC animated:YES];
}

@end

File: Examples/Conversations/Swift/ConversationsExample/QuestionsViewController.swift
//
//  QuestionsViewController.swift
//  Conversations
//
//  Copyright © 2016 Bazaarvoice. All rights reserved.
//

import UIKit
import BVSDK

class QuestionsViewController: UIViewController, UITableViewDataSource, UITableViewDelegate  {
  
  @IBOutlet weak var questionsTableView : BVQuestionsTableView!
  var questions : [BVQuestion] = []
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    questionsTableView.dataSource = self
    questionsTableView.delegate = self
    questionsTableView.estimatedRowHeight = 80
    questionsTableView.rowHeight = UITableViewAutomaticDimension
    questionsTableView.register(UINib(nibName: "MyQuestionTableViewCell", bundle: nil), forCellReuseIdentifier: "MyQuestionTableViewCell")
    
    let questionsRequest = BVQuestionsAndAnswersRequest(productId: Constants.TEST_PRODUCT_ID, limit: 20, offset: 0)
    
    // optionally add in a sort option
    questionsRequest.sort(by: .questionSubmissionTime, monotonicSortOrderValue: .ascending)
    // optionally add in a filter
    questionsRequest.filter(on: .questionHasAnswers, relationalFilterOperatorValue: .equalTo, value: "true")
    
    self.questionsTableView.load(questionsRequest, success: { (response) in
      
      self.questions = response.results
      self.questionsTableView.reloadData()
      
    }) { (error) in
      
      print(error)
      
    }
  }
  
  // MARK: UITableViewDatasource
  
  func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
    return "Question Responses"
  }
  
  func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return questions.count
  }
  
  func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = tableView.dequeueReusableCell(withIdentifier: "MyQuestionTableViewCell")! as! MyQuestionTableViewCell
    
    cell.question = questions[indexPath.row];
    
    return cell
  }
  
  // MARK: UITableViewDelegate
  
  func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
    // Get the answers for this question
    let question = questions[indexPath.row]
    
    let storyboard = UIStoryboard(name: "Main", bundle: nil)
    let answersVC = storyboard.instantiateViewController(withIdentifier: "AnswersViewController") as! AnswersViewController
    answersVC.question = question
    self.navigationController?.pushViewController(answersVC, animated: true)
  }
}
